Realm/ObjectStore/object_store.cpp

```diff
@@ -0,0 +1,418 @@
+
+            // add any new properties (new name or different type)
+            if (!current_prop || property_has_changed(*current_prop, target_prop)) {
+                switch (target_prop.type) {
+                        // for objects and arrays, we have to specify target table
+                    case PropertyTypeObject:
+                    case PropertyTypeArray: {
+                        realm::TableRef link_table = ObjectStore::table_for_object_type(group, target_prop.object_type);
+                        target_prop.table_column = table->add_column_link(realm::DataType(target_prop.type), target_prop.name, *link_table);
+                        break;
+                    }
+                    default:
+                        target_prop.table_column = table->add_column(realm::DataType(target_prop.type), target_prop.name);
+                        break;
+                }
+                changed = true;
+            }
+        }
+
+        // remove extra columns
+        sort(begin(current_schema.properties), end(current_schema.properties), [](Property &i, Property &j) {
+            return j.table_column < i.table_column;
+        });
+        for (auto& current_prop : current_schema.properties) {
+            auto target_prop = target_object_schema->property_for_name(current_prop.name);
+            if (!target_prop || property_has_changed(current_prop, *target_prop)) {
+                table->remove_column(current_prop.table_column);
+                changed = true;
+            }
+        }
+
+        // update table metadata
+        if (target_object_schema->primary_key.length()) {
+            // if there is a primary key set, check if it is the same as the old key
+            if (current_schema.primary_key != target_object_schema->primary_key) {
+                set_primary_key_for_object(group, target_object_schema->name, target_object_schema->primary_key);
+                changed = true;
+            }
+        }
+        else if (current_schema.primary_key.length()) {
+            // there is no primary key, so if there was one nil out
+            set_primary_key_for_object(group, target_object_schema->name, "");
+            changed = true;
+        }
+    }
+
+    // remove primary key entries for deleted types
+    // FIXME - delete actual tables once we have proper testing
+    ObjectStore::Schema schema;
+    for (int i = (int)group->size() - 1; i >= 0 ; i--) {
+        string object_type = object_type_for_table_name(group->get_table_name(i));
+        if (object_type.length()) {
+            if (target_type_names.find(object_type) == target_type_names.end()) {
+                set_primary_key_for_object(group, object_type, "");
+            }
+        }
+    }
+
+    return changed;
+}
+
+bool ObjectStore::is_migration_required(realm::Group *group, uint64_t new_version) {
+    uint64_t old_version = get_schema_version(group);
+    if (old_version > new_version && old_version != NotVersioned) {
```

Comment:
Is this enough to detect a schema mismatch? I would assume that you need to validate all tables as a user might accidentally set the same version even though tables has changed?